### PR TITLE
DTSPO-14865-replace-kured

### PR DESCRIPTION
**Pull request by JoshuaGlassKainos**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DTSPO-14865

### Change description ###
Updated cluster file to include node_os_channel_upgrade 
Updated variables to accommodate node os upgrade flag

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: 01-inputs-required.tf
Comment (reespozzi) [inline 01-inputs-required.tf:76]:
```suggestion
variable "node_os_channel_upgrade" {
```

I think makes more sense if you'll be using the value passed in to set this, it can still default to false and otherwise use the passed in value

Comment (JoshuaGlassKainos) [reply 01-inputs-required.tf:76]: Sounds good, will make the changes here

File: 12-kubernetes-cluster.tf
Comment (MartyFox) [inline 12-kubernetes-cluster.tf:117]: Reading the TF docs:
``` 
 node_os_channel_upgrade must be set to NodeImage if automatic_channel_upgrade has been set to node-image
 ```

I dont think a dynamic block is required here as its a single input value.
Check out this modules example: https://github.com/Azure/terraform-azurerm-aks/blob/main/main.tf#L605C1-L608C6

Also worth looking at the preconditions in that module.

I think you only need but you might need to figure out the relationship between `automatic_channel_upgrade` and `node_os_channel_upgrade`:
```
node_os_channel_upgrade =  var.node_os_channel_upgrade ? "NodeImage" : null
```

Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:117]: Thanks Marty, will read into it further and update the changes needed.

Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:117]: Had a look into it further the auto_channel_upgrade is hardcoded to patch [```automatic_channel_upgrade = var.enable_automatic_channel_upgrade_patch == true ? "patch" : null```](https://github.com/hmcts/aks-module-kubernetes/blob/481054ea9fe1e320db942591cec9a85928a13ae2/12-kubernetes-cluster.tf#L137) and considering the variable is boolean this cannot be used in the precondition to check if it is node-image. 

For the moment I think we only need ```node_os_channel_upgrade =  var.node_os_channel_upgrade ? "NodeImage" : null``` tho case could be said to change ```null``` to ```none``` as the docs here specify https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#node_os_channel_upgrade

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [inline 12-kubernetes-cluster.tf:115]:
```suggestion
  }
```

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [reply 12-kubernetes-cluster.tf]: I think `null` may work best, https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-node-os-image#register-nodeosupgradechannelpreview -- I think if not using dynamic blocks, you'll have to add the feature to all clusters before merging this if you used `None`?

Did a tfplan against some other env show any intended changes with this set to null?

Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf]: Yeah I think if ```None``` was used all clusters will need to have this feature set and ```null``` will circumvent that. 

Did a test in sbox env and the node_os_upgrade showed as ```NodeImage``` and with demo env the node_os_upgrade did not appear. So far it only applies to the sbox env only

**Review by reespozzi — APPROVED** [review]
LGTM, not sure if @MartyFox had any other suggestions

**Comment by JoshuaGlassKainos** [comment]
> LGTM, not sure if @MartyFox had any other suggestions

Will hold off on merge until @MartyFox is good with the suggested changes also.

**Comment by JoshuaGlassKainos** [comment]
Had a chat with Marty, looked into the relationship between `automaic_channel_upgrade` set to `patch` and `node_os_channel_upgrade` set to `NodeImage`. there is no link between the two, the only link present would be if `automaic_channel_upgrade` set to `node-image`

**Pull request merged**
